src/combinatorics/double_counting.lean

```diff
@@ -79,5 +85,17 @@ lemma card_mul_eq_card_mul [Π a b, decidable (r a b)]
 (card_mul_le_card_mul _ (λ a ha, (hm a ha).ge) $ λ b hb, (hn b hb).le).antisymm $
   card_mul_le_card_mul' _ (λ a ha, (hn a ha).ge) $ λ b hb, (hm b hb).le
 
+lemma card_le_card_of_forall_subsingleton [Π a b, decidable (r a b)]
+  (hs : ∀ a ∈ s, ∃ b, b ∈ t ∧ r a b) (ht : ∀ b ∈ t, ({a ∈ s | r a b} : set α).subsingleton) :
```

Comment:
I'm wondering if we should have some more API for these hypotheses (not necessarily as a prerequisite for this PR, unless the right notions already exist in mathlib somewhere, if you'd kindly investigate this).

The first is that `r` is "cosurjective on `s` and `t`," and the second is that `r` is "injective on `s` and `t`." For example, a relation is called a function if it is cosurjective and coinjective.

Reply:
These are [`relator.left_total`](https://leanprover-community.github.io/mathlib_docs/logic/relator.html#relator.left_total) and [`relator.right_unique`](https://leanprover-community.github.io/mathlib_docs/logic/relator.html#relator.right_unique).

I am explicitly not using them here because I want those lemma to be easily discoverable, and their proof is already somewhat of a stretch (if I told that each element of `s` corresponds to an element of `t` and no two elements of `t` correspond to the same element of `s`, would you think double counting?).

Reply:
I think you should use those definitions anyway, and add a docstring to make things more discoverable

Reply:
Ah, I now see that they would operate on the subtype. Can you add those lemmas as well then?